.github/workflows/CICD.yaml

```diff
@@ -8,53 +8,264 @@ on:
+          If "false" CICD will not rewrite the BL_Python dependencies.
+        type: choice
+        options:
+          - true
+          - false
+        default: "true"
+        required: true
+      use_dependency_cache:
+        description:
+          If "true", the workflow will use the dependency cache if the associated
+          cache key has a hit. If "false," any existing dependency cache will be ignored.
+          The workflow will save a new dependency cache with the same cache key regardless
+          of `use_dependency_cache` if there was not a cache hit.
+        type: choice
+        options:
+          - true
+          - false
+        default: "true"
+        required: false
 
 jobs:
-  All:
-    name: Build and test BL_Python
+  Checkout:
+    name: Checkout and Setup
     runs-on: ubuntu-latest
-    strategy:
-      matrix:
-        python-version: ["3.10"]
+
+    outputs:
+      cache-key-dependencies: ${{ steps.generate-cache-keys.outputs.cache_key }}
+      cache-key-run: ${{ steps.generate-cache-keys.outputs.cache_key_run }}
+
+    env:
+      PYTHON_VERSION: "3.10"
 
     steps:
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
+
+      - name: Generate cache keys
+        id: generate-cache-keys
+        # Use separate cache keys for dependencies and for run workspace.
+        # The dependencies can be restored between runs,
+        # while the workspace is restored between jobs in the same run.
+        run: |
+          rand=$(( (1000+$RANDOM)%$RANDOM*$RANDOM ))
+          run_attempt=$([ -n "${{ github.run_attempt }}" ] && printf ${{ github.run_attempt }} || printf $rand)
+          cache_key_dependencies='${{ runner.os }}-dependencies-${{ hashFiles('pyproject.toml', 'src/*/pyproject.toml') }}'
+          cache_key_run='${{ runner.os }}-run-${{ hashFiles('pyproject.toml', 'src/*/pyproject.toml') }}-run-id_${{ github.run_id }}-number_${{ github.run_number }}-attempt_'"$run_attempt"
+          echo "cache_key_dependencies=$cache_key_dependencies" >> $GITHUB_OUTPUT
+          echo "cache_key_run=$cache_key_run" >> $GITHUB_OUTPUT
+
+      #- name: Restore dependency cache
+      #  if: ${{ success() && (github.event_name != 'workflow_dispatch' || inputs.use_dependency_cache == 'true') }}
+      # TODO figure out a way to get 3rd party dependencies cached without caching BL_Python ... maybe.
+      #  uses: actions/cache/restore@v4
+      #  id: restore-dependency-cache
+      #  with:
+      #    path: |
+      #      .github-venv
+      #      node_modules
+      #    key: ${{ steps.generate-cache-keys.outputs.cache_key_dependencies }}
 
-      - name: Set up Python ${{ matrix.python-version }}
-        uses: actions/setup-python@v4
+      - name: Set up Python ${{ env.PYTHON_VERSION }}
+        #if: ${{ success() && (inputs.use_dependency_cache == 'false' || !steps.restore-dependency-cache.outputs.cache-hit) }}
```

Comment:
Do you want to leave this here?

Reply:
For now, yes. It's part of the commented TODOs around caching.